### ISSUE TITLE
Blueprint and WebSockets

### ISSUE THREAD

**Issue by bonigarcia** [apiaryio/api-blueprint#257]
Is it possible to use api-blueprint to document an API based on WebSockets (instead of REST)?

Thanks!

**Comment by zdne** [comment]
Hey @bonigarcia,  
 unfortunately no, not at the moment. There is no dedicated support for WebSocket in the API Blueprint. You can still, however, document its WebSockets HTTP handshake. 

The API Blueprint as it stands right now is closely tied to HTTP protocol (not REST architecture per se). 

If you have any ideas regarding WebSockets support we will be happy to hear them! 

Thanks!

**Comment by bonigarcia** [comment]
Thank for your reply @zdne!

Well, I suppose an smooth integration of Apiary and WebSockets would be by means of the metadata `HOST`. If a WebSocket URI is defined instead of HTTP URL, Apiary could perform a WebSocket connection to this URI, exchanging messages as defined by the specification.

    HOST: ws://host:port/path

**Comment by bonigarcia** [comment]
I am trying to document the WebSocket handshake but I am not able to do it. My WebSocket URI is the following:

    ws://kurento.lab.fiware.org:8888/kurento

The Blueprint code I am trying is as follows:

    FORMAT: 1A
    HOST: http://kurento.lab.fiware.org:8888

    # Stream-oriented

    The Stream Oriented API is a resource-oriented API accessed via WebSockets that uses JSON-RPC V2.0 based representations for information exchange.

    ## WebSocket Handshake [/kurento]

    To establish a WebSocket connection, the client sends a WebSocket handshake request, for which the server returns a WebSocket handshake response.

    ### WebSocket Handshake [GET]

    + Request

        + Headers
    
                Upgrade: websocket
                Connection: Upgrade
                Sec-WebSocket-Key: daecIhfhHeDfYP4rJivWjQ==
                Sec-WebSocket-Version: 13
                Pragma: no-cache
                Cache-Control: no-cache

    + Response 101

And I get the following response:

    We are sorry, but the API call failed. Please, give it another try.

Maybe the problem here is the HTTP 101 response (Switching Protocols) which is returned in the WebSocket handshake.

Does api-blueprint support this kind of HTTP response?

**Comment by zdne** [comment]
@bonigarcia yes API Blueprint should support this. Where are you getting the problem? Using Apiary mock?

**Comment by akoroljov** [comment]
Hi everybody,

I want share with you my own (and my friends and collegues) vision on websocket feature.

WebSocket itself is not alternative for REST of course. It's different things in nature. We saw already existing modules and plugins, which implement REST on top of WebSocket. But it's all about stateful communications and it is not what we are thinking of and not what we really need.

So the question I guess is not only about websocket, but about something that can be "alternative" for REST. Maybe "alternative" is not correct word. But something that can be used instead of REST. Another/additional way to communicate.


What I want to see in specification is support for stateless communications.
How it will look like in test API environment (IMHO and will be great  to discuss it together to understand if its true or not):
	- Specify endpoint. Path is optional. It can be simple IP-address, or even existing websocket connection pool (whatever). But to simplify for the first, we can use the same like URI. You need to think of it like it happens not only in browser. It can be mobile or even something else if it exists.
	- When websocket connection is opened, we can send and receive messages now
	- Sending and receiving messages happens asynchronous and a bit different from REST. For example, in REST, when client sends GET request, then he waits for response from server for this GET request. If connection if timed out, for example, and client does not get response from the server, then client needs to resend this GET request again. But it is not true for "websocket". In websocket it can happen like that: when connection is dropped (timed out), client reopens websocket connection and receive response from the server, because server is already knows that client didn't received this response yet.
	- So the next rule is: client can receive messages even without sending any requests. It can be some status messages, chat messages, or whatever you want.
	- This is full duplex connection. So it’s two way communication. Client can send messages and receive response for it. Can send messages and not wait for response (unbelievable situation, but why not). Client can wait for messages from server and not send any request to server. When it receive message from server, then it can send back to server response message.
	- Websocket connection is alive for the whole runtime lifecycle of the application. This is the main goal in most scenarios
	- We have new keyword - "message". It is not request or response. It is message. And we need also mechanism to answer with messages to some messages, if we need it, in any direction.

This is how I see this new feature. Will be great to discuss it and find the way to implement it.
I am 100% ready to work on it. So my own human/developer resources are shared with you.

With best regards,
Artem

**Comment by akoroljov** [comment]
this is good article to understand the difference
https://blogs.oracle.com/PavelBucek/entry/websocket_vs_rest

and understand that it is not really replacement of REST, but simply another technology to communicate

**Comment by akoroljov** [comment]
WSDL - can be the real answer. Blueprint already support WADL (REST nature). so point may be to extend blueprint to support WSDL.
What do you think about it?
